src/NumericalIntegration.cpp

```diff
@@ -0,0 +1,367 @@
+                                                       const Vector3d& wD, const Vector3d& vD,
+                                                       double step, double prec = 1e-10)
+{
+  auto qf = rbd::SO3Integration(qi, wi, wD, step, prec, prec);
+  if (!qf.second)
+  {
+    return freeJointIntegration_(qi, xi, qf.first, wi, vi, wD, vD, step, prec);
+  }
+  else
+  {
+    double halfStep = step / 2;
+    auto H = freeJointIntegration(qi, xi, wi, vi, wD, vD, halfStep, prec);
+    return freeJointIntegration(H.first, H.second, wi + halfStep * wD, vi + halfStep * vD, wD, vD, halfStep, prec);
+  }
+}
+
+Quaterniond sphericalJointIntegration(const Quaterniond& qi, const Vector3d& wi,
+                                      const Vector3d& wD, double step, double prec = 1e-10)
+{
+  auto qf = rbd::SO3Integration(qi, wi, wD, step, prec, prec, true);
+  if (!qf.second)
+  {
+    return qf.first;
+  }
+  else
+  {
+    double halfStep = step / 2;
+    auto q = sphericalJointIntegration(qi, wi, wD, halfStep, prec);
+    return sphericalJointIntegration(q, wi + halfStep * wD, wD, halfStep, prec);
+  }
+}
+
+Matrix2d R(double q)
+{
+  double c = std::cos(q);
+  double s = std::sin(q);
+  return (Matrix2d() << c, -s, s, c).finished();
+}
+
+std::pair<double, Vector2d> planarJointIntegration(double qi, const Vector2d& xi,
+                                                   double wi, Vector2d vi,
+                                                   double wD, const Vector2d& vD,
+                                                   double step, double prec = 1e-10)
+{
+  double erri = fourthDerivativeSquaredNormPlanar(vi, wi, vD, wD);
+  double errf = fourthDerivativeSquaredNormPlanar(vi + step * vD, wi + step * wD, vD, wD);
+  double errMax = std::sqrt(std::max(erri, errf));
+  int n = static_cast<int>(std::ceil(step / 2 * std::sqrt(std::sqrt(errMax * step / (180 * prec)))));
+  if(n == 0) n = 1;
+
+  double nthStep = step / n;
+  double halfNthStep = nthStep / 2;
+  double nthStep2 = nthStep * halfNthStep;
+  double halfNthStep2 = halfNthStep * halfNthStep / 2;
+
+  std::pair<double, Vector2d> H = {qi, R(qi) * (6 * xi + nthStep * vi)};
+  for(int i = 0; i < n - 1; ++i)
+  {
+    Vector2d vh = vi + vD * halfNthStep;
+    Vector2d ve = vi + vD * nthStep;
+    double qh = H.first + wi * halfNthStep + wD * halfNthStep2;
+    double qe = H.first + wi * nthStep + wD * nthStep2;
+    H.second += nthStep * (4 * R(qh) * vh + 2 * R(qe) * ve);
+    H.first = qe;
+    vi = ve;
+    wi += nthStep * wD;
+  }
+  Vector2d vh = vi + vD * halfNthStep;
+  Vector2d vf = vi + vD * nthStep;
+  double qh = H.first + wi * halfNthStep + wD * halfNthStep2;
+
+  H.first += wi * nthStep + wD * nthStep2;
+  H.second += nthStep * (4 * R(qh) * vh + R(H.first) * vf);
+
+  H.second = R(H.first).transpose()*(H.second / 6);
+  //H.second /= 6;
+
+  return H;
+}
+
+} // namespace
+
+namespace rbd
+{
+
+std::pair<Quaterniond, bool> SO3Integration(const Quaterniond & qi,
```

Comment:
Is c++17 ok? If so, `std::optional` would suit exactly the behavior you want with the pairs.